### PR TITLE
Rewrite Promise constructor description

### LINKED ISSUE #7034: Issue with "Promise() constructor": various statements are false/incorrect, and others are incomplete/misleading/nonsensical
MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/Promise

#### What information was incorrect, unhelpful, or incomplete?

Quote: "The operation within executor is asynchronous and provides a callback."

This is false: the executor is executed synchronously.

Quote: "The callback is defined within the executor code."

What does this even mean? The the executor *is* the callback (passed to `new Promise`).

Quote: "The callback terminates by invoking resolutionFunc."

This is false: the callback continues executing after that function call.

Quote: "The invocation of resolutionFunc includes a value parameter."

This is incomplete: This value parameter is optional

Quote: "The value is passed back to the tethered Promise object."

The word "tethered" should not be used at all. It is non-standard. The executor just calls back.

Quote: "The Promise object (asynchronously) invokes any associated .then(handleResolved)"

This is misleading, as it does not invoke `.then`, but `handleResolved`.

Quote: "The value received by .then(handleResolved) is passed to the invocation of handleResolved as an input parameter"

This makes no sense. The value passed to `handleResolved` is the value the original promise resolved with, i.e. the value passed to `resolutionFunc`.

#### Specific section or headline?

Parameters - executor

#### What did you expect to see?

Information in line with ECMAScript specification. Content that has the high quality I am used to when reading MDN. This is a rare exception.

#### Did you test this? If so, how?

n/a

<details>
<summary>MDN Content page report details</summary>

* Folder: `en-us/web/javascript/reference/global_objects/promise/promise`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/Promise
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/javascript/reference/global_objects/promise/promise/index.html
* Last commit: https://github.com/mdn/content/commit/75bf139c604bdca9e955c32618787a31458512b2
* Document last modified: 2021-06-23T08:11:11.000Z

</details>

### PR DESCRIPTION
**Pull request by Josh-Cena**
#### Motivation

Fixes #7034. Most of the sentiments in that issue are misunderstanding of the document (for example, "callback" means the callback-based API, not `executor`), but we can benefit from some more clarity.

#### Metadata

- [ ] Adds a new document
- [x] Rewrites (or significantly expands) a document
- [ ] Fixes a typo, bug, or other error

### CODE REVIEW

File: files/en-us/web/javascript/reference/global_objects/promise/promise/index.md [hunk 1]
```diff
@@ -25,57 +24,87 @@ new Promise(executor)
 ### Parameters
 
 - `executor`
+  - : A {{jsxref("function")}} to be executed by the constructor. It receives two functions as parameters: `resolutionFunc` and `rejectionFunc`. Any errors thrown in the `executor` will cause the promise to reject, and the return value is neglected. The semantics of `executor` are explained in detail below.
```
Comment (MrBrain295) [inline files/en-us/web/javascript/reference/global_objects/promise/promise/index.md:27]:
```suggestion
  - : A {{jsxref("function")}} to be executed by the constructor. It receives two functions as parameters: `resolutionFunc` and `rejectionFunc`. Any errors thrown in the `executor` will cause the promise to be rejected, and the return value will be neglected. The semantics of `executor` are detailed below
```

File: files/en-us/web/javascript/reference/global_objects/promise/promise/index.md [hunk 2]
```diff
@@ -25,57 +24,87 @@ new Promise(executor)
...
 
-    `resolutionFunc` and `rejectionFunc` are also functions, and you can give them whatever actual names you want. Their signatures are simple: they accept a single parameter of any type.
+## Description
 
-    <!-- prettier-ignore -->
-    ```js
-        resolutionFunc(value) // call on resolved
-        rejectionFunc(reason) // call on rejected
-    ```
+Traditionally (before promises), asynchronous tasks are designed as callbacks.
```
Comment (MrBrain295) [inline files/en-us/web/javascript/reference/global_objects/promise/promise/index.md:35]:
```suggestion
Traditionally (before promises), asynchronous tasks were designed as callbacks.
```

File: files/en-us/web/javascript/reference/global_objects/promise/promise/index.md [hunk 3]
```diff
@@ -25,57 +24,87 @@ new Promise(executor)
...
-    5. The `value` is passed back to the tethered `Promise` object.
-    6. The `Promise` object (asynchronously) invokes any associated `.then(handleResolved)`.
-    7. The `value` received by `.then(handleResolved)` is passed to the invocation of `handleResolved` as an input parameter (see [Chained Promises](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#chained_promises)).
+```js
+resolutionFunc(value) // call on resolved
+rejectionFunc(reason) // call on rejected
+```
 
-### Return value
+The `resolutionFunc` `value` parameter can be another promise object, in which case the promise gets dynamically inserted into the [promise chain](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#chained_promises). The `rejectionFunc` has semantics close to the [`throw`](en-US/docs/Web/JavaScript/Reference/Statements/throw) statement, so `reason` is typically an [`Error`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error) instance. If `value` or `reason` is omitted, the promise is resolved/rejected with `undefined`.
```
Comment (MrBrain295) [inline files/en-us/web/javascript/reference/global_objects/promise/promise/index.md:68]:
```suggestion
The `resolutionFunc` `value` parameter can be another promise object, in which case the promise gets dynamically inserted into the [promise chain](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#chained_promises). The `rejectionFunc` has semantics close to the [`throw`](en-US/docs/Web/JavaScript/Reference/Statements/throw) statement, so `reason` is typically an [`Error`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error) instance. If either the `value` or the `reason` are omitted, the promise is resolved/rejected with `undefined.`
```

Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/global_objects/promise/promise/index.md:68]: either...or is used as singular.

**Review by teoli2003 — CHANGES_REQUESTED** [review]

File: files/en-us/web/javascript/reference/global_objects/promise/promise/index.md [hunk 4]
```diff
@@ -25,57 +24,87 @@ new Promise(executor)
...
-    is custom code that ties an outcome to a promise. You, the programmer, write the
-    `executor`. Its signature is expected to be:
+### Return value
 
-    ```js
-        function(resolutionFunc, rejectionFunc){
-          // typically, some asynchronous operation.
-        }
-    ```
+When called via `new`, the `Promise` constructor returns a promise object. The promise object will become "resolved" when either of the functions `resolutionFunc` or `rejectionFunc` are invoked. Note that if you call `resolutionFunc` or `rejectionFunc` and pass another Promise object as an argument, it can be said to be "resolved", but still not "settled".
```
[the remark this answers is not in the document]
Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/global_objects/promise/promise/index.md:31]: There's no place on top of my mind (but I don't typically read those guide pages unless someone reports an issue😬). The `Promise.resolve()` page is also very unclear on this (and someone has already complained in https://github.com/mdn/content/issues/334) so I'll probably eventually write one somewhere.

**Review by teoli2003 — APPROVED** [review]
👍

**Comment by teoli2003** [comment]
Yes, we probably need this terminology explained somewhere, maybe there: https://developer.mozilla.org/en-US/docs/Glossary/Promise.

**Pull request merged**